### PR TITLE
sqlstore: use xorm.EngineGroup and add database.slaves setting

### PR DESCRIPTION
**Pull request by Thib17**

### CODE REVIEW

File: conf/defaults.ini [hunk 1]
```diff
@@ -68,6 +68,7 @@ socket = /tmp/grafana.sock
 # Either "mysql", "postgres" or "sqlite3", it's your choice
 type = sqlite3
 host = 127.0.0.1:3306
+slaves =
```
Comment (iksaif) [inline conf/defaults.ini]: Add something here: https://github.com/grafana/grafana/blob/fd5fc0b7abc2058e61d7214634a1a4eaa1bc7ee8/docs/sources/installation/configuration.md#host

Comment (Thib17) [reply conf/defaults.ini]: Done

**Comment by Thib17** [comment]
cc @bergquist

File: pkg/services/sqlstore/quota.go [hunk 2]
```diff
@@ -98,15 +99,17 @@ func UpdateOrgQuota(cmd *m.UpdateOrgQuotaCmd) error {
 	return inTransaction(func(sess *DBSession) error {
 		//Check if quota is already defined in the DB
 		quota := m.Quota{
-			Target: cmd.Target,
-			OrgId:  cmd.OrgId,
+			Target:  cmd.Target,
```
Comment (bergquist) [inline pkg/services/sqlstore/quota.go]: Are these changes related to the rest of the PR? seems odd :)

File: pkg/services/sqlstore/sqlstore.go [hunk 3]
```diff
@@ -26,6 +27,7 @@ import (
 
 type DatabaseConfig struct {
 	Type, Host, Name, User, Pwd, Path, SslMode string
+	Slaves                                     []string
```
Comment (bergquist) [inline pkg/services/sqlstore/sqlstore.go]: Its time we stop using the `master/slave` terminology and use `primary/replica` instead.

File: pkg/services/sqlstore/quota.go (hunk 2 above)
Comment (Thib17) [reply pkg/services/sqlstore/quota.go]: As explained in the commit message there is 2 side effects. One of them is that zero time is now translated to nil before sql insertion. This raise errors in tests that were not previously raised: `'NOT NULL constraint failed: dashboard_acl.created'`.
From my point of view, error are rightly raised since created/updated fields are sometimes never filled.
So I filled such (missing ?) fields for `quota` and `dashboard_acl`.

File: pkg/services/sqlstore/sqlstore.go (hunk 3 above)
Comment (Thib17) [reply pkg/services/sqlstore/sqlstore.go]: Done

**Comment by bergquist** [comment]
Another idea to solve kinda of the same problem would be to introduce caching in the ORM mapper. http://gobook.io/read/github.com/go-xorm/manual-en-US/chapter-11/#

**Comment by iksaif** [comment]
> solve kinda of the same problem

Which problem? I suspect that you are thinking about performance. We don't really have replicas for performance, we have them for redundancy (in order to make them master if the master fails). But since they are here already, and sometimes in a dc that is closer to the master dc, we might as well use them for reads too :)

The cache, in a multiple instance setup, will probably be not super easy to configure correctly.

**Comment by bergquist** [comment]
Oh. My mistake. Yea I assumed you wanted it for performance reasons :)

**Comment by iksaif** [comment]
@Thib17  can you rebase ?

**Comment by torkelo** [comment]
@DanCech Any thoughts on this?  Adding support for replicas (master / slave) setup for databases.  A feature in xorm, https://github.com/go-xorm/xorm/issues/78

I am a little bit hesistant to start using this complex xorm feature as I do not fully trust the messy xorm code base.

**Comment by iksaif** [comment]
Unless you have a MySQL cluster with a very fast failover, having this would allow to have something that mostly work while failing over to the new master.

This also make reads faster in multi-dc deployments, which is nice :).

(But if there is another way to get the same thing, I'm fine too)

**Comment by torkelo** [comment]
Thanks for contributing and wanting to make Grafana better! Sad to report that I will be closing this. 

Only for now until we have a feature request and clear demand for a feature like this.
